Update portal

TimotheCity opened the pull request.

TimotheCity committed:
handle shortcode property to override css & call setBtnCss

TimotheCity committed:
Add css override to btn style

TimotheCity committed:
rework css to only need one shortcode attribute for user

TimotheCity committed:
get css from shortcode & check roughly css validity

TimotheCity committed:
fix indentation

TimotheCity committed:
css

TimotheCity committed:
flags

TimotheCity committed:
js

TimotheCity committed:
translations

TimotheCity committed:
Update login button to send form data on connect

TimotheCity committed:
Add user form setting page

TimotheCity committed:
Add user form in login page

TimotheCity committed:
sanitize form inputs

TimotheCity committed:
Create default form value, language

TimotheCity committed:
Add script for login page

TimotheCity committed:
Add user form and lang selection to login page

TimotheCity committed:
Add script to form setting

TimotheCity committed:
Add form setting to menu

TimotheCity committed:
Update form language

TimotheCity committed:
Handle form create & update field

TimotheCity committed:
Create user info datatable

TimotheCity committed:
Merge branch 'shortcode-css-injection-for-login-button' into update-portal

TimotheCity committed:
fix css

TimotheCity committed:
fix regex

GroupeCitypassengerInc commented on admin/digilan-admin.php:
```diff
@@ -417,6 +424,119 @@ private static function updateCityscopeCloud($cityscopeCloud)
+    private static function add_field_to_form()
+    {
+        $fields = get_option('user_form_fields');
+        $new_field_data = array_reduce(
+            array_keys($_POST),
+            function($acc, $post_key) {
+                [$prefix, $field_option, $lang] = explode('/', $post_key);
+                if ($prefix !== 'digilan-token-new-field') {
+                    return $acc;
+                }
+                $value = DigilanTokenSanitize::sanitize_post($post_key);
+                if ($lang) {
+                    $acc[$field_option][$lang] = $value;
+                } else {
+                    $acc[$field_option] = $value;
+                }
+                return $acc;
+            },
+            array(),
+        );
```
Please use a loop here , it s php , it can do that but usual reader will have trouble this javascript like 
also, it s kinda bad because the function is recompile completly each time

TimotheCity replied:
Just a classic `foreach` loop ?
`foreach ($_POST as $key -> $value);`

dohnuts replied:
c est toujours un array_reduce

GroupeCitypassengerInc commented on admin/digilan-admin.php:
```diff
@@ -417,6 +424,119 @@ private static function updateCityscopeCloud($cityscopeCloud)
+            self::add_field_to_form();
+        }
+        if (isset($_POST['digilan-token-user_form_fields'])) {
+            self::update_user_form_fields();
```
put the exist here not inside the sub routine for clarity

GroupeCitypassengerInc commented on admin/digilan-admin.php:
```diff
@@ -417,6 +424,119 @@ private static function updateCityscopeCloud($cityscopeCloud)
+    private static function updated_fields_position(&$current_form_field, $key, &$pos_counter) {
+        $current_form_field['position'] = $pos_counter;
+        $pos_counter++;
+    }
```
Unsure why you have position inside an ARRAY, it s already positional.

Maybe INSERTING / MOVING the element at the right position instead of managing a index would be way more clean ( and easier to read the code )

GroupeCitypassengerInc commented on admin/digilan-admin.php:
```diff
@@ -577,6 +697,38 @@ public static function test_url_backend()
+            exit();
+        }
+
+        $form_languages = get_option('form_languages');
```
If that is the get_option of the wp api the KEY must be prefixed and more precise 

This assume ALL the PORTAL share the same conf ?

GroupeCitypassengerInc commented on admin/digilan-admin.php:
```diff
@@ -577,6 +697,38 @@ public static function test_url_backend()
+        $form_languages = get_option('form_languages');
+        $lang_code = $form_languages[$lang]['code'];
+        $user_id = get_current_user_id();
+        update_user_meta($user_id,'user_lang',$lang_code);
```
https://developer.wordpress.org/reference/functions/get_current_user_id/ or ZERO if not logged 

it s never a good practice to ignore returns and documentation ( get_option can fail to )

GroupeCitypassengerInc commented on admin/digilan-admin.php:
```diff
@@ -577,6 +697,38 @@ public static function test_url_backend()
+        update_user_meta($user_id,'user_lang',$lang_code);
+    }
+
+    public static function update_form_language()
+    {
```
any reason to perform sanitize, get_option databases read and the only update_user_meta or only update option, ? one function doing both would be enough no ?

TimotheCity replied:
I have :
- a function to allow admin to add or remove languages of custom portal 
- a function to allow visitor to select dispalyed language of custom portal

dohnuts replied:
Nomme les clairements alors, et check les droits avant de les appeler

GroupeCitypassengerInc commented on digilan-token.php:
```diff
@@ -159,6 +162,99 @@ public static function init()
+                    'fr_FR' => 'Français, English, Español',
+                ),
+                'position'     => 3,
+            ),
+            'stay-length' => array(
+                'display-name' => array(
+                    'en_US' => 'Stay length',
+                    'fr_FR' => 'Durée du séjour',
+                ),
+                'instruction'  => array(
+                    'en_US' => 'Stay length in days',
+                    'fr_FR' => 'Durée du séjour en jours',
+                ),
+                'type'         => 'number',
+                'unit'         => array(
+                    'en_US' => 'days',
+                    'fr_FR' => 'jours',
+                ),
+                'position'     => 4,
+            )
+        );
+        add_option('user_form_fields', $user_form_fields);
+        add_option('typeOptions', array(
+            'text'     => 'Text',
+            'email'    => 'Email',
+            'tel'      => 'Tel',
+            'number'   => 'Number',
+            'radio'    => 'Radio buttons',
+            'select'   => 'Drop-down menu',
+            'checkbox' => 'Checkbox',
+        ));
+        add_option('form_languages', array(
+            'English'    => array('name' => 'English'   , 'frenchName' => 'Anglais'   , 'code' => 'en_US', 'implemented' => true ),
+            'French'     => array('name' => 'French'    , 'frenchName' => 'Français'  , 'code' => 'fr_FR', 'implemented' => true ),
+            'German'     => array('name' => 'German'    , 'frenchName' => 'Allemand'  , 'code' => 'de_DE', 'implemented' => false),
+            'Italian'    => array('name' => 'Italian'   , 'frenchName' => 'Italien'   , 'code' => 'it_IT', 'implemented' => false),
+            'Portuguese' => array('name' => 'Portuguese', 'frenchName' => 'Portuguais', 'code' => 'pt_PT', 'implemented' => false),
+            'Spanish'    => array('name' => 'Spanish'   , 'frenchName' => 'Espagnol'  , 'code' => 'es_ES', 'implemented' => false),
+        ));
+
```
so yeah there s offical rule about plugins in wordpress and naming policy , all the option need to be named more precisely

dohnuts replied:
Peut on mettre ces machin la dans un fichier à part et le lire ? ( aussi les clef des add_option sont pas prefixé correctement )

GroupeCitypassengerInc commented on digilan-token.php:
```diff
@@ -1142,7 +1285,20 @@ public static function authenticate_ap_user_on_wp($username = '')
+                    $acc[$field_key] = $field_value;
+                    return $acc;
+                },
+            );
```
for loop and explicit die on sanitize failed

GroupeCitypassengerInc replied:
also sanitize should now the type of data , this code is not very clear, is it actually testing the type and value coming from the user ?

GroupeCitypassengerInc commented on digilan-token.php:
```diff
@@ -1277,6 +1433,24 @@ public static function digilantoken_add_query_vars($vars)
+    public static function search_for_lang_by_code($code, $langs)
+    {
+        $user_lang_key = array_search($code, array_column($langs, 'code', 'name'));
+        return $user_lang_key ?? array_keys($langs)[0];
```
cute ?? but again not easy to read, moreover this default should not be first in a default but a get_option('dlt_default_lang')  or the default language of the all wordpress

GroupeCitypassengerInc commented on digilan-token.php:
```diff
@@ -1277,6 +1433,24 @@ public static function digilantoken_add_query_vars($vars)
+    {
+        $user_lang_code = get_user_meta( get_current_user_id(), 'user_lang', true) ?? get_user_locale();
+        $form_languages = get_option("form_languages");
+
```
' everywhere but there "  and naming is wrong

GroupeCitypassengerInc commented on digilan-token.php:
```diff
@@ -1277,6 +1433,24 @@ public static function digilantoken_add_query_vars($vars)
+        $user_lang_code = get_user_meta( get_current_user_id(), 'user_lang', true) ?? get_user_locale();
+        $form_languages = get_option("form_languages");
+
+        $form_languages_implemented = array_filter($form_languages, function($lang) {
+            return $lang['implemented'];
+        });
+        $user_lang_key = self::search_for_lang_by_code($user_lang_code, $form_languages_implemented);
+
```
search_for_lang_by_code could be a for loop and have the array

this is just making the serveur work for nothing ( go over array to create a new one, then go over the new array to find a key ... ) 

and it does not provide any hint to a reader ( especially a non multi language experienced one )

GroupeCitypassengerInc commented on includes/digilan-db.php:
```diff
@@ -131,6 +131,18 @@ static function wp_digilan_social_users() {
+                `user_id` INT,
+                `user_info` VARCHAR(1024),
+                FOREIGN KEY `fk_%sdigilan_token_meta_1` (user_id) REFERENCES %sdigilan_token_users_1(id)
+            );",
```
i know it s incorrect in other part of code but if possble use the version instead of _1

GroupeCitypassengerInc commented on includes/digilan-db.php:
```diff
@@ -131,6 +131,18 @@ static function wp_digilan_social_users() {
+        return sprintf(
+            "CREATE TABLE %sdigilan_token_meta_users_1 (
+                `user_id` INT,
+                `user_info` VARCHAR(1024),
```
why 1024 ?

GroupeCitypassengerInc replied:
this limit does not appear anywhere elese

TimotheCity replied:
I don't know the limite
I can put VARCHAR(MAX) but this has a little impact on performance

GroupeCitypassengerInc replied:
https://mariadb.com/kb/en/text/

1/ will you do search on it ?
2/ will you do indexed search on it ?
3/ can you store in a more structured way to have limited filed and unlimited field

Nobody knows the limit; they are changing over time.

check what already exists 
```
describe wp_options;
+--------------+---------------------+------+-----+---------+----------------+
| Field        | Type                | Null | Key | Default | Extra          |
+--------------+---------------------+------+-----+---------+----------------+
| option_id    | bigint(20) unsigned | NO   | PRI | NULL    | auto_increment |
| option_name  | varchar(191)        | NO   | UNI |         |                |
| option_value | longtext            | NO   |     | NULL    |                |
| autoload     | varchar(20)         | NO   | MUL | yes     |                |
+--------------+---------------------+------+-----+---------+----------------+
```

Justify static choice in a comment.

Profit.

GroupeCitypassengerInc commented on includes/digilan-sanitize.php:
```diff
@@ -268,6 +327,37 @@ public static function sanitize_get($in)
+            if ($prefix === 'dlt-user-form-hidden') {
+                $user_form_fields = get_option('user_form_fields');
+                if ($unsafe_value === '') {
+                    return $unsafe_value;
```
so this is the real lsanitize but when called earlier it s hidden , and why calll a global sanitize_get then if case in it ?

=> create a sanitize dlt-user-form-hidden and make code readable 

it s more custom_form_portal than dlt-user-form- ( all form are used by a user ) ... you re working on the portal_form to make it custom.

GroupeCitypassengerInc commented on includes/digilan-user-form.php:
```diff
@@ -0,0 +1,159 @@
+    {
+        $res = '';
+        foreach ($user_form_fields_in as $key => $value)
+        {
+            $res .= '<input type="hidden" name="dlt-user-form-hidden/' . $key .'" value="">';
+        }
```
pretty sure any php coder is screaming in pain here , check ob_start , the fact that stdout is going out and how to do a template in php ?

dohnuts replied:
resolved ou ?

GroupeCitypassengerInc commented on includes/digilan-user-form.php:
```diff
@@ -0,0 +1,159 @@
+                <img
+                    class="language-flag"
+                    src="'. plugins_url($src, DLT_ADMIN_PATH) .'"
+                    alt="'. $lang["name"] .' flag"
```
please check other code and documentation , it s really hard to read ( you did not evuse a multiline var with << )

GroupeCitypassengerInc commented on providers/mail/mail.php:
```diff
@@ -45,19 +45,39 @@ public function getOptionKey()
+                $acc[$field_key] = $field_value;
+                return $acc;
+            },
+            array(),
```
same comment use procedural style

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+(function ($) {
+  $(document).ready(function () {
+    $('.flag').hover(function() {
+      let delete_button = this.nextElementSibling;
+      $(delete_button).css('display', 'block');
+    }, function() {
+      let delete_button = this.nextElementSibling;
+      $(delete_button).css('display', 'none');
+    });
+    $('.lang-flag-delete').hover(function() {
+      $(this).css('display', 'block');
+    }, function() {
+      $(this).css('display', 'none');
+    });
```
Pourquoi cela n'est pas fait en css ?

```css
.flag:hover + button {
  display: block;
}

.lang-flag-delete {
  display: block;
}
```

TimotheCity replied:
Pour que je m'entraine en jQuerry :)

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+
+    $('#lang-search').on('input', function(input) {
+      let search = input.target.value.toLowerCase().trim();
+      var listItems = $("#language_list li");
```
var

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+      listItems.each(function(idx, li) {
+        let name = $(li).attr('name').toLowerCase();
+        if (name.includes(search)) {
+          $(li).css('display', '');
```
Invalid display. Block ? Inline block ?

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+
+    function update_language(lang) {
+      $.ajax({
+        type: 'post',
```
POST en majuscule

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+        data: {
+          '_ajax_nonce': user_form_data._ajax_nonce,
+          'action': 'digilan-token-form-language-settings',
+          'lang': lang,
```
Pas de ''

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+        dataType: 'json',
+        url: ajaxurl,
+        success: function () {
+          location.reload();
```
C'est barbare, tu peux pas l'ajouter dynamiquement en js ?

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+    });
+    $('.lang-select').focusout(function () {
+      // Timeout required, otherwise, pop-up list close before item click handled
+      setTimeout(() => {
+        $('.language_list_container').hide();
+       }, 100);
```
Il y'a partout des fonctions sans le raccourcis () => {...}, il faut rester consistant

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+    });
+
+    $('.form-settings-field-row').on('click', function(value) {
+      currentElement = value.target;
+      sibling = currentElement.nextElementSibling;
```
Définition des variables ? let ?

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+        return;
+      }
+
+      while (!sibling || !sibling.className.includes('edit-form-settings-field-row')) {
+        currentElement = currentElement.parentElement;
+        sibling = currentElement.nextElementSibling;
+      };
```
Je sais pas ce que c'est que ca mais ca me semble très gitant. + Pas de ; après un while

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+        $(currentElement).addClass('top-row-visible');
+        $(currentElement.nextElementSibling).addClass('bottom-row-visible');
+      };
+    });
+    $('.delete-field').on('click', function (value) {
```
Saut de ligne entre deux fonctions ce serait mieux ? Ca vaut partout

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+      };
+    });
+    $('.delete-field').on('click', function (value) {
+      let isChecked = value.target.checked
```
;

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+    });
+    $('.update-field').on('input', function (value) {
+      let row = value.target.closest("div[name='field-row']");
+      let resetButton = $(row).find("input[name='reset-changes-button']")[0];
```
Fait une vérification qu'il y'a bien qq chose avant de pick [0]

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+
+      while (isNoChanges && i < fields.length) {
+        [prefix, property, field_name, lang] = fields[i].name.split('/');
+        isNoChanges = fields[i].value === user_form_fields[field_name][property][lang];
```
Ca sort d'ou ce truc ? user_form_fields

TimotheCity replied:
Dans `digilan-token.pop` (L550) :
`wp_localize_script` permet d'utiliser des variables du php depuis js

```            
$user_form_fields = get_option('user_form_fields');
wp_localize_script('dlt-user-form-fields', 'user_form_fields', $user_form_fields);
```

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+        [prefix, property, field_name, lang] = fields[i].name.split('/');
+        isNoChanges = fields[i].value === user_form_fields[field_name][property][lang];
+        i++;
+      }
```
for

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+        isNoChanges = fields[i].value === user_form_fields[field_name][property][lang];
+        i++;
+      }
+      if (isNoChanges) {
```
Saut de ligne avant

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+    $('input[name="reset-changes-button"]').on('click', function(value) {
+      let row = value.target.closest("div[name='field-row']");
+      $(row).removeClass('update-in-progress');
+      $(value.target).attr('disabled', true);
```
${this} ?

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+      let i = 0;
+
+      while (isNoChanges && i < fields.length) {
+        [prefix, property, field_name, lang] = fields[i].name.split('/');
```
let ? Il faut mettre un mot avant de définir des variables 😳

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+
+      let fields = $(row).find("input[type='text']");
+      for (let i = 0; i < fields.length; i++) {
+        [prefix, property, field_name, lang] = fields[i].name.split('/');
```
let ? Il faut mettre un mot avant de définir des variables 😳

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+      $('label[name="instruction"]').css('display', 'block');
+
+      $('label[name="options"]').css('display', 'none');
+      $('label[name="options"]').children('.required_input').removeAttr("required");
```
Créer une variable puis appel 
```js
let options = $('label[name="options"]');
options.css('display', 'none');
options.children('.required_input').removeAttr("required");
```

ou chain

```js
$('label[name="options"]')
  .css('display', 'none')
  .children('.required_input')
  .removeAttr("required");
```

Mais la il y'a deux recherche dans le DOM pour la même chose

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+          $('label[name="unit"]').css('display', 'block');
+          break;
+        case 'radio':
+          $('label[name="options"]').css('display', 'block');
+          $('label[name="options"]').children('.required_input').attr("required", true);
```
Met ca dans une fonction et appelle si radio ou select

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+      $('#multiple').css('display', 'none');
+
+      let type = value.target.name;
+      document.getElementById("new-field-type").value = type;
```
Attention à l'utilisage des doubles quotes. Essaye d'être consistent. Cela vaut partout.

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+    });
+    $('#copy-shortcode').on('click', function() {
+      let copyText = document.getElementById("form-shortcode");
+      let copyButton = document.getElementById("copy-shortcode");
```
$(this) ?

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+      }
+    });
+    $('#copy-shortcode').on('click', function() {
+      let copyText = document.getElementById("form-shortcode");
```
$(this).first() ou $(this).next() ?

De manière générale il vaut mieux essayer de multiplier les appels DOM trop couteux

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+      const buttonWidth = copyButton.offsetWidth;
+      copyButton.style.width = `${buttonWidth}px`;
+      copyButton.value = js_translation.copied_shortcode;
+      copyText.select();
+      copyText.setSelectionRange(0, 99999);
```
Comment tu peux le select et mettre une range si c'est un text. Si c'est une input le nom de variable n'est pas bon.

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+      navigator.clipboard.writeText(copyText.value);
+      copyText.blur();
+
+      $(this).addClass('success');
```
Parfois this , parfois copyButton 😛

nathangobinet commented on js/user-form.js:
```diff
@@ -0,0 +1,58 @@
+(function($) {
+  $(document).ready(function() {
+    $('.lang-select').click(function () {
+      $('.language_list_container').show();
+    });
+    $('.lang-select').focusout(function () {
+      // Timeout required, otherwise, pop-up list close before item click handled
+      setTimeout(() => {
+        $('.language_list_container').hide();
+       }, 100);
+    });
```
Pourquoi il y'a une duplication de code entre les deux fichiers ?

TimotheCity replied:
Les scripts sont localisé:
`user-form.js` est utilisé sur la page du formulaire
`user-form-settings.js` sur la page de settings

Pour ne pas dupliquer le code, je peux créer un fichier avec juste cette fonction et `wp_register_script` sur les 2 pages

nathangobinet commented on js/user-form.js:
```diff
@@ -0,0 +1,58 @@
+        },
+      });
+    };
+    $('#language_list li').click(function(){
```
Cela ne devrait pas être possible de cliquer sur des li. Il y'a un problème d'accessibilité. Uniquement les liens et les boutons peuvent être cliquer.

nathangobinet commented on js/user-form.js:
```diff
@@ -0,0 +1,58 @@
+        dataType: 'json',
+        url: '/wordpress/wp-admin/admin-ajax.php',
+        success: function () {
+          location.reload();
```
Comme précedemment

nathangobinet commented on js/user-form.js:
```diff
@@ -0,0 +1,58 @@
+      update_form_display_language(lang)
+    });
+
+    $('.missing_translation').each(function(){
+      $(this).prop('title', 'Missing translation');
+    })
```
Pourquoi cela est fait en javascript ?

TimotheCity replied:
Pour éviter de l'écrire partout en phtml, je sais pas si c'était judicieux

nathangobinet commented on js/user-form.js:
```diff
@@ -0,0 +1,58 @@
+    function update_data_form(key, input_value) {
+      $('input[name="dlt-user-form-hidden/' + key + '"]').val(input_value);
+      $('a[name="connection-link-form"]').attr('href', function(i, a)
+      {
```
Saut de ligne inutile

nathangobinet commented on js/user-form.js:
```diff
@@ -0,0 +1,58 @@
+      });
+    };
+    jQuery.each(form_inputs, function(key, value) {
+      _name = 'dlt-' + key ;
```
_name ?
Pourquoi un undescore ?
Pourquoi pas de let ?

nathangobinet commented on js/user-form.js:
```diff
@@ -0,0 +1,58 @@
+    };
+    jQuery.each(form_inputs, function(key, value) {
+      _name = 'dlt-' + key ;
+      $('input[name="' + _name + '"]').on("change", function updated_data() {
```
Pourquoi utliser une fonction non anonyme ?

nathangobinet commented on js/admin/user-form-settings.js:
```diff
@@ -0,0 +1,190 @@
+      setTimeout(() => {
+        $(this).removeClass('success');
+        copyButton.value = js_translation.copy_shortcode_button;
+      }, "2000");
```
chaine de caractère à la place d'un number ?

TimotheCity committed:
Use css:hover instead of jquerry

TimotheCity committed:
Use let not var

TimotheCity committed:
Fix empty display property

TimotheCity committed:
fix "POST" instead of "post"

TimotheCity committed:
Remove ' from ajax data keys and replace keys with "-" to "_"

TimotheCity committed:
Replace arrow function for consistency

TimotheCity committed:
Use let on variables

TimotheCity committed:
Readability

TimotheCity committed:
Missing ;

TimotheCity committed:
Use for loop instead of while

TimotheCity committed:
Use `this` instead of `value.target` when possible

TimotheCity committed:
Remove deadcode

TimotheCity committed:
"let" is missing

TimotheCity committed:
chain function instead of searching DOM 2 times

TimotheCity committed:
Quote / double quote consistency

TimotheCity committed:
use this and fix button feedback

TimotheCity committed:
useless line break

TimotheCity committed:
Rework name and deadcode

TimotheCity committed:
Lighten cost to retrieve element

TimotheCity committed:
Duplicate code in function

TimotheCity committed:
use anonymous function

TimotheCity committed:
Remove useless code

TimotheCity committed:
use - instead of _ in classname

TimotheCity committed:
use button and clean css

nathangobinet reviewed (approved):
JS ok ✅

TimotheCity commented on admin/digilan-admin.php:
```diff
@@ -156,6 +159,8 @@ public static function admin_init()
         add_action('admin_post_digilan-token-plugin', 'DigilanTokenAdmin::save_form_data');
         add_action('wp_ajax_digilan-token-plugin', 'DigilanTokenAdmin::ajax_save_form_data');
         add_action('wp_ajax_digilan-token-cityscope', 'DigilanTokenAdmin::test_url_backend');
+        add_action('wp_ajax_digilan-token-form-language-settings', 'DigilanTokenAdmin::update_form_language');
```
kjdzakjdza

TimotheCity committed:
Improve clarity

TimotheCity committed:
Remove unused position

TimotheCity committed:
Add prefixe to options

TimotheCity committed:
array_reduce is bad practice

TimotheCity committed:
Check function returned value

TimotheCity committed:
die on sanitize false

TimotheCity committed:
Readability

TimotheCity committed:
uncomited changes, fix

TimotheCity committed:
Invert condition check

TimotheCity committed:
Simplify loop

TimotheCity committed:
Fix datatable version numbers

TimotheCity committed:
simplify sanitize

TimotheCity committed:
rename dlt-user-form to custom-form-portal

TimotheCity committed:
use ob_start

TimotheCity committed:
Clarity

TimotheCity committed:
Justify user_info VARCHAR size

TimotheCity committed:
Improve readability

TimotheCity committed:
Rename stuff for clarity

dohnuts commented on admin/digilan-admin.php:
```diff
@@ -417,6 +424,112 @@ private static function updateCityscopeCloud($cityscopeCloud)
+            add_option("digilan_token_user_form_fields", array());
+        }
+        $new_field_data = array();
+        foreach($_POST as $post_key => $value) {
+            [$prefix, $field_option, $lang] = explode('/', $post_key);
+            if ($prefix !== 'digilan-token-new-field') {
+                continue;
+            }
+
+            $value = DigilanTokenSanitize::sanitize_post_form_fields($post_key);
+            if (false === $value) {
+                \DLT\Notices::addError(sprintf('Invalid value for %s', $field_option));
+                wp_redirect(self::getAdminUrl('form-settings'));
+                exit();
+            }
+            if ($lang) {
+                $new_field_data[$field_option][$lang] = $value;
+            } else {
+                $new_field_data[$field_option] = $value;
+            }
+        }
```
C est pas une bonne idée de faire cela ici
IE faire un truc automatique pour de la verif
Dans la sous fonction tu as encore le switch case qui a 5 case donc ici 
tu as plus de ligne que de cas

$lang = $lang || 'en' ( utilise une langue default pour simplifier tout au lieu d avoir dedans ou un sous truc qui fait des types mixer qui vont casser le cerveau, 'en' ou meme 'default' )
$new_field_data[$field_option][$lang] = DigilanTokenSanitize::sanitize_post_form_EMAIL_fields($_POST['trucmail']);

dohnuts commented on admin/digilan-admin.php:
```diff
@@ -417,6 +424,112 @@ private static function updateCityscopeCloud($cityscopeCloud)
+            if ($prefix !== 'form-fields') {
+                continue;
+            }
+            if (in_array($field_name, $deleted_keys)) {
+                continue;
+            }
+            if ($field_option == 'delete') {
+                array_push($deleted_keys, $field_name);
+                unset($user_form_fields[$field_name]);
+                continue;
+            }
+
+            $value = DigilanTokenSanitize::sanitize_post_form_fields($post_key);
+            if (false === $value) {
+                \DLT\Notices::addError(sprintf('Invalid %s for %s', $field_option, $field_name));
+                wp_redirect(self::getAdminUrl('form-settings'));
+                exit();
+            }
+            $input_type = $user_form_fields[$field_name]['type'];
```
pareil

dohnuts commented on admin/digilan-admin.php:
```diff
@@ -577,6 +690,47 @@ public static function test_url_backend()
+            wp_die('There is no languages','fatal');
+        }
+        $lang_code = $form_languages[$lang]['code'];
+        $user_id = get_current_user_id();
+        if ($user_id === 0) {
+            error_log('User is not logged');
+        }
```
en general on commence par la... aussi comment un utilisateur non logger est arrivé jusqu a update_user_language() ???? 

Ce serait bien d avoir un peu de contexte aussi ( c est quoi check_ajax_referer )

TimotheCity replied:
`check_ajax_referer` est utilisé à d'autres endroit dans le plugin
Cela permet de vérifier que la requête ne provient pas d'un autre site

dohnuts commented on admin/digilan-admin.php:
```diff
@@ -577,6 +690,47 @@ public static function test_url_backend()
+    {
+        check_ajax_referer('digilan-token-form-language-settings');
+        $lang = DigilanTokenSanitize::sanitize_post('custom_portal_lang');
+        if (false === $lang) {
```
Je prefere un sanitize_custom_lang($_value) c est mieux ( pour un tas de raison )

Apres si tu veux garder le concept de faire grep -r '_POST' et verif que tout est dans DigilanTokenSanitize tu faire  DigilanTokenSanitize::sanitize_custom_lang();

( pareil au dessus ) mais evite le on method avec des if c est juste moche ( old code )

dohnuts commented on admin/digilan-admin.php:
```diff
@@ -577,6 +690,47 @@ public static function test_url_backend()
+        }
+        $current = $form_languages[$lang]['implemented'];
+        $form_languages[$lang]['implemented'] = !$current;
+        update_option('digilan_token_form_languages', $form_languages);
```
dans les languages modernes y a des jolies invert 
la doc city de codage dit que ! est pas lisible ( ici tu fait un invert et c est le sook amr ? ) donc ( cf pr )

dohnuts committed:
Commenting to make the single char clear

According to science `https://stackoverflow.com/questions/4603589/boolean-value-switch-invert`

```
Time to swap values 1 millon times:

$bool ^= 1 -> 0.2595 seconds
$bool = $bool!=true -> 0.2803 seconds
$bool = !$bool -> 0.2833 seconds
$bool = ($bool-1)*(-1) -> 0.3047 seconds
$bool = $bool?0:1 -> 0.3253 seconds
if -> 0.3589 seconds
switch -> 0.4387 seconds
$bool = abs($bool-1) -> 1.2211 seconds
```

mais c est surtout pour eviter un '$current' et un '!$current'

dohnuts commented on admin/templates/form-settings.php:
```diff
@@ -0,0 +1,392 @@
+      <img
+        class="language-flag"
+        src="<?php echo plugins_url($src, DLT_ADMIN_PATH) ?>"
+        alt="<?= $lang['name'] ?> flag"
```
`<?= plugins_url($src, DLT_ADMIN_PATH) ?>`  comment les autres ???
je sais plus si en php y a aussi <?: pour encoder/raw .. à voir

dohnuts commented on admin/templates/form-settings.php:
```diff
@@ -0,0 +1,392 @@
+            style="width:100%"
+            <?php if( $is_required ): ?>
+              required
+            <?php endif; ?>
```
`<?= ($is_required) ? "required" ?>`  le plus clair de mémoire

dohnuts commented on admin/templates/form-settings.php:
```diff
@@ -0,0 +1,392 @@
+      </fieldset>
+    </td>
+  </tr>
+  <?php return ob_get_contents();
```
va la ligne ici ( comme partout ailleurs, cela montre bien que c'est la fin )
<?php
return ob_get_contents();


mais perso je ferais le ob_start / ob_get_content en dehors de la fonction ( et la fonction serait dansun fichie php a part , je suppose vu que c est un template) 

genre ```php
ob_start;
require template_file_new_field_lang_row.php;
ob_get_contents();
```

dohnuts commented on admin/templates/form-settings.php:
```diff
@@ -0,0 +1,392 @@
+
+$types = get_option('digilan_token_type_options_display_name');
+defined('ABSPATH') || die();?>
+<div class="dlt-admin-content">
```
si possible aller a la ligne apres le die pour bien voir le debut et pas de \n entre ?> et <div

dohnuts replied:
et pas de \n entre ?> et <div

dohnuts commented on admin/templates/form-settings.php:
```diff
@@ -0,0 +1,392 @@
+$types = get_option('digilan_token_type_options_display_name');
+defined('ABSPATH') || die();?>
+<div class="dlt-admin-content">
+  <h1><?php _e('Form configuration', 'digilan-token'); ?></h1>
```
un bon moyen de faire mieux que les if required au dessus pareil que _e

dohnuts commented on digilan-token.php:
```diff
@@ -1277,6 +1453,39 @@ public static function digilantoken_add_query_vars($vars)
+        $user_id = get_current_user_id();
+        if ($user_id === 0) {
+            return current($form_languages);
+        }
```
pareil en premier verifier le login, apres la db, et c est pas normal qu'un utilisateur non loggé arrive la ! comment est ce possible ?

TimotheCity replied:
Tu m'as dit de check les valeurs de retour, en effet ce n'est pas possible
`https://developer.wordpress.org/reference/functions/get_current_user_id/ or ZERO if not logged`
`it s never a good practice to ignore returns and documentation ( get_option can fail to )`

dohnuts reviewed (changes requested):
J'espère que cela fait du sens.

TimotheCity committed:
Merge pull request #2 from dohnuts/patch-1

Commenting to make the single char clear

TimotheCity committed:
close php on new line

TimotheCity committed:
open php should have a line for itself

TimotheCity committed:
Clean one line ternary condition

TimotheCity committed:
code homogeneity

TimotheCity committed:
Rename

TimotheCity committed:
Revert "Commenting to make the single char clear"

TimotheCity committed:
Sanitize_lang

TimotheCity committed:
Duplicated code in function

TimotheCity committed:
Fix flip boolean value

dohnuts committed:
With the right value

Keep the comment !

TimotheCity committed:
Move constant options in constant file

TimotheCity committed:
cleaner way

TimotheCity committed:
Merge pull request #4 from dohnuts/patch-2

With the right value

TimotheCity committed:
Merge pull request #3 from TimotheCity/revert-2-patch-1

Revert "Commenting to make the single char clear"

TimotheCity committed:
Merge branch 'update-portal' of https://github.com/TimotheCity/digilan-token into update-portal

TimotheCity committed:
rework sanitize

TimotheCity committed:
Fix update

TimotheCity committed:
wp_get_current_user instead of get current id

dohnuts commented on admin/digilan-admin.php:
```diff
@@ -417,6 +424,135 @@ private static function updateCityscopeCloud($cityscopeCloud)
+                return DigilanTokenSanitize::sanitize_form_field_options($unsafe_value);
+                break;
+            default:
+                _default_wp_die_handler(sprintf('Unhandled field option: %s', $field_key));
+                break;
```
je pense qu il faut appeler wp_die ici pas hooker vers le default

dohnuts commented on admin/digilan-admin.php:
```diff
@@ -417,6 +424,135 @@ private static function updateCityscopeCloud($cityscopeCloud)
+            add_option("digilan_token_user_form_fields", array());
+        }
+        $new_field_data = array();
+        foreach($_POST as $post_key => $post_value) {
+            // $post_key example : 
+            // - dlt-custom-portal-new-field/display-name/fr_FR
```
pas d iteration sur _POST ! merci

dohnuts commented on admin/digilan-admin.php:
```diff
@@ -417,6 +424,135 @@ private static function updateCityscopeCloud($cityscopeCloud)
+            }
+        }
+        $new_field_data_filtered = array_filter($new_field_data, function($data) {
+            return is_array($data) ? array_filter($data) : $data;
+        });
```
array_filter
avec un seul argument ?
`If no callback is supplied, all empty entries of array will be removed. See [empty()](https://www.php.net/manual/en/function.empty.php) for how PHP defines empty in this case.``

donc tu essaie d enlever des null dans des sous tableaux ?

Ces sous tableaux existerais pas si tu avais suivi mes reco précédente: 

langue par default et pas d iteration.

Tout cela est compliqué pour pas grand chose et dur à valider. 

Par ailleurs , les commentaires au dessus aide pas du tout 
`// handle hidden field options without translations (like input type)`

oui on avait compris que y avais pas de langue.... mais cela handle rien

par contre quand tu écris 
``` 
 $new_field_data_filtered = array_filter($new_field_data, function($data) {
            return is_array($data) ? array_filter($data) : $data;
        });
```
pour 
// delete empty value in subarrays

en faisant 4 appel externe sur deux lignes :-/ pour des truc que tu viens de créer c est vraiment tres tres moyen

encore une fois l iteration est pas une bonne idée

Ta liste de post devrait etre plus structuré tu recois genre la liste des champs dans un truc , tu verifies sui leur nom est valide puis tu itere la dessus pour extraire leurs infos.

Aussi des structures dynamiques à type variable c'est trop compliqué , créer la langue DEFAULT ou choisie une langue default et drop le cas pas de langue merci

dohnuts commented on admin/digilan-admin.php:
```diff
@@ -417,6 +424,135 @@ private static function updateCityscopeCloud($cityscopeCloud)
+            return is_array($data) ? array_filter($data) : $data;
+        });
+
+        $first_translation_name = current($new_field_data_filtered['display-name']);
+        $new_field_key = str_replace(' ', '-', strtolower($first_translation_name));
+        if ($user_form_fields[$new_field_key]) {
+            \DLT\Notices::addError(__('Field name already exist', 'digilan-token'));
```
Je suppose qu il vaudrait mieux interdires des espaces ,
en tout cas la il faut justifier avec un commentaires pourquoi un espace est enlever ,
en HTML tu peux encoder en (htmlencode ) donc space est %20 des attributs 

Tu peux double check avec @nathangobinet  et la spec html 

Bref à justifier, encoder ou plutôt virer

dohnuts commented on admin/digilan-admin.php:
```diff
@@ -417,6 +424,135 @@ private static function updateCityscopeCloud($cityscopeCloud)
+        }
+        $deleted_keys = array();
+
+        foreach ($_POST as $post_key => $post_value) {
+            // $post_key example :
```
non (*3)

TimotheCity committed:
Fix sanitize empty

dohnuts commented on admin/digilan-admin.php:
```diff
@@ -577,6 +713,44 @@ public static function test_url_backend()
+    public static function update_form_language()
+    {
+        check_ajax_referer('digilan-token-form-language-settings');
+        $lang = DigilanTokenSanitize::sanitize_custom_lang($_POST['custom_portal_lang']);
+        if (false === $lang) {
+            \DLT\Notices::addError(sprintf(__('Form can\'t be translated in %s.'), $lang));
```
GOOD

dohnuts commented on admin/templates/form-settings.php:
```diff
@@ -0,0 +1,388 @@
+            id="new-field-options-<?= $lang['code'] ?>"
+            style="width:100%"
+            <?php // use class for jquery to handle "required" with "display:none" input ?>
+            <?= ($is_required) ? 'class="required_input"' : '' ?>"
+            />
```
fun fact si dns is_required tu met 'class="required_input"' ou rien au lieu de fair eun boolean ( et tu l apelle input_option ) tu peux faire plus et l info est proche de la décision

dohnuts replied:
je vois pas de résolution la

dohnuts commented on digilan-token.php:
```diff
@@ -1141,8 +1225,44 @@ public static function authenticate_ap_user_on_wp($username = '')
+        $user_info = array();
         if ($user_id == false) {
-            DigilanTokenUser::create_ap_user($mac, $social_id);
+            foreach ($_GET as $get_key -> $get_value) {
+                [$prefix, $field_key] = explode('/', $get_key);
+                if ($prefix !== 'custom-form-portal-hidden') {
+                    continue;
+                }
+
+                $field_value;
+                switch ($field_key) {
+                    case 'text':
+                        $field_value = DigilanTokenSanitize::sanitize_custom_form_portal_hidden_text($get_value);
+                        break;
+                    case 'number':
+                        $field_value = DigilanTokenSanitize::sanitize_custom_form_portal_hidden_number($get_value);
+                        break;
+                    case 'email':
+                        $field_value = DigilanTokenSanitize::sanitize_custom_form_portal_hidden_email($get_value);
+                        break;
+                    case 'tel':
+                        $field_value = DigilanTokenSanitize::sanitize_custom_form_portal_hidden_tel($get_value);
+                        break;
```
for each inconu do if ( somethingX )  do somethingY
vers
try_do_somethingY(somethingX )

merci ( pas de for pour faire jolie, des fonctions pour faire claire )

dohnuts reviewed (changes requested):
J ai pas tout review trop de comment

TimotheCity committed:
Fix hidden inputs name

TimotheCity committed:
Update the way to sanitize user inputs :
now loop on existing fields instead of POST_key

dohnuts commented on admin/digilan-admin.php:
```diff
@@ -503,30 +512,37 @@ private static function update_user_form_fields()
-                exit();
+            $field_type = $form_field_value['type'];
+            foreach($form_languages as $lang) {
+                if (false === $lang['implemented']) {
+                    continue;
+                }
+                $lang_code = $lang['code'];
+                $safe_display_name = self::sanitize_form_field_option_by_type('display-name', $_POST["form-fields/display-name/$form_field_key/$lang_code"]);
+                $user_form_fields[$form_field_key]['display-name'][$lang_code] = $safe_display_name;
+                
+                $safe_instruction = self::sanitize_form_field_option_by_type('instruction', $_POST["form-fields/instruction/$form_field_key/$lang_code"]);
+                $user_form_fields[$form_field_key]['instruction'][$lang_code] = $safe_instruction;
+                
+                if ($field_type === 'radio' || $field_type === 'select') {
+                    $safe_options = self::sanitize_form_field_option_by_type('options', $_POST["form-fields/options/$form_field_key/$lang_code"]);
+                    $user_form_fields[$form_field_key]['options'][$lang_code] = $safe_options;
+                }
+                
+                if ($field_type === 'number') {
+                    $safe_unit= self::sanitize_form_field_option_by_type('unit', $_POST["form-fields/unit/$form_field_key/$lang_code"]);
+                    $user_form_fields[$form_field_key]['unit'][$lang_code] = $safe_unit;
+                }
```
all of that goes into $user_form_fields[$form_field_key] = routine

self::sanitize_form_field_option_by_type('display-name' cela sert a rien de fair eun switch case dedans apres tu factorise du vide

et lis ton post dedans

aussi un nom de fonction peut etre une variable...

TimotheCity committed:
Simplify string concat

TimotheCity committed:
Subroutine for field translations creation & update

TimotheCity committed:
Rename language functions

TimotheCity committed:
Comment field key creation

dohnuts commented on digilan-token.php:
```diff
@@ -814,11 +865,18 @@ public static function widgetShortcode($atts)
             'fontsize' => 16
         ), $atts);
 
+        $override_btn_css = $atts["override-btn-css"];
+        $css_regex = '/^([A-Za-z\-]+?:\s*?[0-9A-Za-z\-%#\s]+?;\s*?)*?$/';
+        preg_match($css_regex, $override_btn_css, $valid_css_input);
```
a mon avis quelque commentaires ferais pas de mal dans le coin

dohnuts commented on digilan-token.php:
```diff
@@ -814,11 +865,18 @@ public static function widgetShortcode($atts)
+
         $providersIn = array();
         foreach (self::$providers as $provider) {
             if ($provider->getState() == 'configured') {
                 $provider_id = $provider->getId();
                 if ($atts[$provider_id] == 1) {
+                    if ($valid_css_input) {
+                        $provider->setBtnCss($override_btn_css);
+                    }
                     $providersIn[$provider_id] = self::$providers[$provider_id];
                 }
             }
```
```php
if ($provider->getState() == 'configured') {
                $provider_id = $provider->getId();
                if ($atts[$provider_id] == 1) {
                    if ($valid_css_input) {
                        $provider->setBtnCss($override_btn_css);
                    }
                    $providersIn[$provider_id] = self::$providers[$provider_id];
                }
            }
```

ce serait pas mal de recoder cela a la mode city 
if ($provider->getState() != 'configured') continue;
$provider_id = $provider->getId();
if ($atts[$provider_id] != 1) continue;

TimotheCity committed:
remove lang argument

dohnuts commented on digilan-token.php:
```diff
@@ -1141,8 +1225,22 @@ public static function authenticate_ap_user_on_wp($username = '')
         if ($user_id == false) {
-            DigilanTokenUser::create_ap_user($mac, $social_id);
+            $user_form_fields = get_option('digilan_token_user_form_fields');
+            foreach($user_form_fields as $form_field_key=>$form_field_value) {
+                $field_type = $form_field_value['type'];
+                if (false === isset($_GET["custom-form-portal-hidden/$field_type/$form_field_key"])) {
+                    continue;
+                }
+                $safe_value = self::sanitize_custom_portal_input($field_type, $_GET["custom-form-portal-hidden/$field_type/$form_field_key"]);
+                if (false === $safe_value) {
+                    continue;
+                }
+                $user_info[$form_field_key] = $safe_value;
```
J voulais accepter mais y a trop de pas ok ici

dohnuts replied:
self::sanitize_custom_portal_input , fait le test isset(arg2) et vire le continue du isset au dessus

Quand un sanitize failed, on continue pas => fait un die dans le sanitize et si le return false est pas un fail
pour le coup tu as pas de early return 

if ($safe_value)  $user_info[$form_field_key] = $safe_value;



et tout le foreach pourrait etre dans 

self::sanitize_custom_portal_inputs($user_info);

dohnuts commented on digilan-token.php:
```diff
@@ -1151,6 +1249,35 @@ public static function authenticate_ap_user_on_wp($username = '')
+            case 'tel':
+                $safe_value = DigilanTokenSanitize::sanitize_custom_form_portal_hidden_tel($unsafe_value);
+                break;
+            case 'radio':
+            case 'select':
+                // selectable values, no sanitize needed
+                $safe_value = $unsafe_value;
```
bas si

dohnuts replied:
https://media.tenor.com/dAjBzKj9H3AAAAAC/danger-alert.gif

dohnuts commented on includes/digilan-user-form.php:
```diff
@@ -0,0 +1,136 @@
+                case 'text':
+                case 'tel':
+                case 'number':
+                case 'email':
+                    [$unit, $unit_class] = self::translate_field($field_data['unit']);
+                    [$display_name, $display_name_class] = self::translate_field($field_data["display-name"]);
+                    [$instruction, $instruction_class] = self::translate_field($field_data["instruction"]);
+                    [$value, $value_class] = self::translate_field($fields_array[$field_key]);
+
+                    $field_label = '<label for="dlt-' . $field_key . '"><strong class="' .$display_name_class .'">' . $display_name . '</strong></label>';
+                    $field_input = ' <input class="regular-text '. $instruction_class .'" type="' . $field_data["type"]  . '" placeholder="' . $instruction . '" name="dlt-' . $field_key .'" '.$field_data['required'] .'>';
+                    $form_inputs .= $field_label . '<div style="display: flex; align-items: center;">'. $field_input . (isset($unit) ? '<span style="margin-left:10px;" class="' . $unit_class . '">' .$unit . '</span>' : "") .'</div>';
+                    break;
+                case 'radio':
+                    [$display_name, $display_name_class] = self::translate_field($field_data["display-name"]);
+                    [$options, $options_class] = self::translate_field($field_data["options"], true);
+
+                    $field_label = '<label for="dlt-' . $field . '"><strong class="' .$display_name_class .'">' . $display_name . '</strong></label>';
+                    $field_radio_buttons = '';
+                    foreach($options as $radioButton) {
+                        $option_input = '<input type="radio" id="' . $radioButton .'" name="dlt-' . $field_key .'" value="' . $radioButton . '" ' . $field_data['required'] .'>';
+                        $option_label = '<label style="margin-left: 5px;" class="' .$options_class .'" for="' . $radioButton . '">' . $radioButton . '</label>';
+                        $field_radio_buttons .='<div>' . $option_input . $option_label . '</div>';
+                    }
+                    $form_inputs = $field_label . '<div style="text-align: left">' . $field_radio_buttons . '</div>';
+                    break;
+                case 'select':
+                    [$display_name, $display_name_class] = self::translate_field($field_data["display-name"]);
+                    [$instruction, $instruction_class] = self::translate_field($field_data["instruction"]);
+                    [$options, $options_class] = self::translate_field($field_data["options"], true);
+
+                    $field_label = '<label for="dlt-' . $field_key . '"><strong class="' .$display_name_class .'">' . $display_name . '</strong></label>';
+                    $field_options = '<option value="" class="' .$instruction_class .'" disabled selected>-- ' . $instruction . ' --</option>';
+                    foreach($options as $option) {
+                        $field_options .= '<option value="'. $option . '"><span class="' .$options_class .'">' . $option . '</span></option>';
+                    }
+                    $field_select = '<select name="dlt-' . $field_key . '" id="' . $field_key . '" '. $field_data['required'] .' style="text-align-last:center;">' . $field_options . '</select>';
+                    $form_inputs .= $field_label . '<div style="display: flex; align-items: center;">' . $field_select . '</div>';
+                    break;
+            }
```
on peut essayer de faire du php SVP ( avec ob ou en direct comme l autre fichier ?)

TimotheCity committed:
additional required class

TimotheCity committed:
Clarity shortcode css test

TimotheCity committed:
return early

TimotheCity committed:
Subfunction sanitize user input

TimotheCity committed:
sanitize options

TimotheCity committed:
ob_start the html

dohnuts commented on admin/templates/form-settings.php:
```diff
@@ -0,0 +1,386 @@
+          style="width: 500px; display: flex;"
+        >
+          <input
+            type="text"
+            name="digilan-token-new-field/unit/<?= $lang['code'] ?>"
+            id="new-field-unit-<?= $lang['code'] ?>"
+            style="width:100%"
+            placeholder="<?php _e('Number unit', 'digilan-token'); ?>"
+            title="<?php _e('Number unit', 'digilan-token'); ?>"
```
pas de pattern

dohnuts commented on admin/templates/form-settings.php:
```diff
@@ -0,0 +1,386 @@
+          style="width: 500px; display: flex;"
+        >
+          <input
+            type="text"
+            placeholder="<?php _e("Options: separate them with a comma [,]", "digilan-token"); ?><?= $input_require_star ?>"
+            title="<?php _e("Options: separate them with a comma [,]", "digilan-token"); ?><?= $input_require_star ?>"
+            name="digilan-token-new-field/options/<?= $lang['code'] ?>"
+            id="new-field-options-<?= $lang['code'] ?>"
+            style="width:100%"
+            <?php // Use class for jquery to handle "required" with "display:none" conflict when options is hidden ?>
```
pas de pattern

dohnuts commented on admin/templates/form-settings.php:
```diff
@@ -0,0 +1,386 @@
+          style="width: 500px; display: flex;"
+        >
+          <input
+            type="text"
+            name="digilan-token-new-field/instruction/<?= $lang['code'] ?>"
+            id="new-field-instruction-<?= $lang['code'] ?>"
+            style="width:100%"
+            placeholder="<?php _e('Instructions', 'digilan-token'); ?><?= $input_require_star ?>"
+            title="<?php _e('Instructions*', 'digilan-token'); ?><?= $input_require_star ?>"
+            <?= ($is_required) ? "required" : '' ?>
```
pas de pattern

dohnuts commented on admin/templates/form-settings.php:
```diff
@@ -0,0 +1,386 @@
+          style="width: 500px; display: flex;"
+        >
+          <input
+            type="text"
+            name="digilan-token-new-field/display-name/<?= $lang['code'] ?>"
+            id="new-field-name-<?= $lang['code'] ?>"
+            placeholder="<?php _e('Field name', 'digilan-token'); ?><?= $input_require_star ?>"
+            title="<?php _e('Field name', 'digilan-token'); ?><?= $input_require_star ?>"
+            style="width:100%"
```
pas de pattern

dohnuts commented on admin/digilan-admin.php:
```diff
@@ -577,6 +702,44 @@ public static function test_url_backend()
+            wp_redirect(self::getAdminUrl('form-settings'));
+            exit();
+        }
+        // flip implemented value
+        $form_languages[$lang]['implemented'] ^= 1;
+        update_option('digilan_token_form_languages', $form_languages);
```
dans quel cas un langugage devient non disponible ?

TimotheCity replied:
J'ai fait en sorte que l'admin ait une liste de langues
Il choisit celles qui seront présente dans le portail
Il doit alors fournir les traductions pour ces champs

Donc une liste n'est pas disponible dans le portail si l'admin ne veut pas l'ajouter à son portail

dohnuts reviewed (changes requested):
presque ok

TimotheCity committed:
Add pattern

dohnuts commented:
Ceci ne reponds pas a la question

regarde le code dans quel cas tu passe dans
$form_languages[$lang]['implemented'] ^= 1; avec
$form_languages[$lang]['implemented'] == 1

?

met donc un log

if ( $form_languages[$lang]['implemented'] == 1 ) {
 error_log ( "deactivating language $lang");
}

et regarde si tu arrive a le faire

On Wed, Nov 30, 2022 at 6:54 AM TimotheCity ***@***.***> wrote:
>
> @TimotheCity commented on this pull request.
>
> ________________________________
>
> In admin/digilan-admin.php:
>
> > +        // flip implemented value
>
> +        $form_languages[$lang]['implemented'] ^= 1;
>
> +        update_option('digilan_token_form_languages', $form_languages);
>
> J'ai fait en sorte que l'admin ait une liste de langues
> Il choisit celles qui seront présente dans le portail
> Il doit alors fournir les traductions pour ces champs
>
> Donc une liste n'est pas disponible dans le portail si l'admin ne veut pas l'ajouter à son portail
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.Message ID: ***@***.***>



-- 
--
---------------------------------------------------------------------------------------------------------------------
Knowing is not enough; we must apply. Willing is not enough; we must do

TimotheCity committed:
ob_start select lang component

TimotheCity committed:
fix clear buffer

TimotheCity committed:
small fix

TimotheCity committed:
Fix binary / boolean value

TimotheCity committed:
Fix empty displayed value

TimotheCity committed:
add error_log
